holmes/plugins/toolsets/coralogix/utils.py

```diff
@@ -0,0 +1,166 @@
+        if add_pod_tag and kubernetes.get("pod_name"):
+            tags.append(f'pod_name="{kubernetes.get("pod_name")}"')
+        if add_namespace_tag and kubernetes.get("namespace_name"):
+            tags.append(f'namespace_name="{kubernetes.get("namespace_name")}"')
+    return " ".join(tags)
+
+
+""" improves human readability of logs by indenting logs that span multiple lines.
+This is important because lines for a single logs() call on a system may span multiple coralogix log lines.
+
+e.g.
+instead of printing this:
+```
+2025-03-25T07:43:39.062945526Z Require stack:
+     at defaultResolveImpl (node:internal/modules/cjs/loader:1061:19)
+     at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1066:22)
+     at Module.require (node:internal/modules/cjs/loader:1491:12)
+   code: 'MODULE_NOT_FOUND',
+2025-03-25T07:43:39.062945626Z etc.
+```
+
+this method will print:
+
+```
+2025-03-25T07:43:39.062945526Z Require stack:
+                                   at defaultResolveImpl (node:internal/modules/cjs/loader:1061:19)
+                                   at resolveForCJSWithHooks (node:internal/modules/cjs/loader:1066:22)
+                                   at Module.require (node:internal/modules/cjs/loader:1491:12)
+                                 code: 'MODULE_NOT_FOUND',
+2025-03-25T07:43:39.062945626Z etc.
+```
+
+
+"""
+
+
+def indent_multiline_log_message(indent_char_count: int, log_message: str):
+    lines = log_message.replace("\r", "\n").split("\n")
+    log_message = lines.pop(0)
+    while (
+        not log_message and len(lines) > 0
+    ):  # Some log messages start with a line feed or return carriage. Make sure the first line is not empty
+        log_message = lines.pop(0)
+    for new_line in lines:
+        if new_line.strip():
+            line = "\n" + (" " * indent_char_count) + new_line
+            log_message += line
+    return log_message
+
+
+def normalize_datetime(date_str: str) -> str:
+    if not date_str:
+        return "UNKNOWN_TIMESTAMP"
+
+    try:
+        date = datetime.fromisoformat(date_str.rstrip("Z"))
+        # Ensure the timestamp has the maximum resolution (nanoseconds)
+        normalized_date_time = date.strftime("%Y-%m-%dT%H:%M:%S.%f")
+        return normalized_date_time + "Z"
+    except Exception:
+        logging.debug(f"Failed to normalize timestamp {date_str}")
+        return date_str
+
+
+def extract_logs(
+    json_objects: List[Dict[str, Any]], add_namespace_tag: bool, add_pod_tag: bool
+) -> str:
+    """Extracts timestamp and log values from parsed JSON objects, sorted in ascending order (oldest first)."""
+    logs = []
+
+    for data in json_objects:
+        if isinstance(data, dict) and "result" in data and "results" in data["result"]:
+            for entry in data["result"]["results"]:
+                try:
+                    user_data = json.loads(entry.get("userData", "{}"))
+                    kubernetes = user_data.get("kubernetes", None)
```

Comment:
Are we sure that those are indeed the generic use case? what is Kubernetes in that case?
log_message will always be in "log" field because in their docs I see there is also `message` field. would we like to do make those configurable?

Reply:
I'm not sure tbh. This is why I made this code flexible and if the `kubernetes` field is not there or has the wrong data the code will still work.
This just adds tags to each log line for Holmes to understand what pod and namespace are the logs from. The value is theoretical and there in case Holmes casts too wide a net and get the logs from many pods.